File: solutions/java/meetup/1/src/main/java/Meetup.java
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.util.function.Predicate;
import java.util.function.Function;
import java.util.Map;
import java.util.EnumMap;

public class Meetup {

    private final LocalDate date;
    private static final Map<MeetupSchedule, Function<DayOfWeek, Predicate<LocalDate>>> clauses = new EnumMap<MeetupSchedule, Function<DayOfWeek, Predicate<LocalDate>>>(MeetupSchedule.class);

    static {
        clauses.put(
            MeetupSchedule.FIRST,
            dayOfWeek -> {
                return day -> day.getDayOfWeek() == dayOfWeek;
            }
        );

        clauses.put(
            MeetupSchedule.SECOND,
            dayOfWeek -> {
                return new Predicate<LocalDate>() {
                    int count = 0;
                    
                    public boolean test(LocalDate date) {
                        if (date.getDayOfWeek() == dayOfWeek) {
                            count++;
                            return count == 2;
                        }
                        return false;
                    }
                };
            }
        );  
        
        clauses.put(
            MeetupSchedule.THIRD,
            dayOfWeek -> {
                return new Predicate<LocalDate>() {
                    int count = 0;
                    
                    public boolean test(LocalDate date) {
                        if (date.getDayOfWeek() == dayOfWeek) {
                            count++;
                            return count == 3;
                        }
                        return false;
                    }
                };
            }
        );  
                
        clauses.put(
            MeetupSchedule.FOURTH,
            dayOfWeek -> {
                return new Predicate<LocalDate>() {
                    int count = 0;
                    
                    public boolean test(LocalDate date) {
                        if (date.getDayOfWeek() == dayOfWeek) {
                            count++;
                            return count == 4;
                        }
                        return false;
                    }
                };
            }
        );  

        clauses.put(
            MeetupSchedule.LAST,
            dayOfWeek -> {
                return day -> day.getDayOfWeek() == dayOfWeek && day.plusDays(7).getMonth() != day.getMonth();
            }
        );

        clauses.put(
            MeetupSchedule.TEENTH,
            dayOfWeek -> {
                return day -> day.getDayOfWeek() == dayOfWeek && day.getDayOfMonth() >= 13 && day.getDayOfMonth() <= 19;
            }
        );
    }

    public Meetup(int monthOfYear, int year) {
        date = LocalDate.of(year, monthOfYear, 1);
    }

    public LocalDate day(DayOfWeek dayOfWeek, MeetupSchedule schedule) {
        Predicate<LocalDate> pred = clauses.get(schedule).apply(dayOfWeek);
        
        for (int i = 0; i < date.getMonth().length(date.isLeapYear()); i++) {
            LocalDate candidate = date.plusDays(i);
            if (pred.test(candidate)) {
                return candidate;
            }
        }
        throw new RuntimeException("Can't find a proper meetup");
    }

}

File: solutions/clojure/meetup/1/src/meetup.clj
(ns meetup
  (:import (java.time YearMonth DayOfWeek)))

(defn leap? [year]
  (and (zero? (mod year 4))
       (or (not= 0 (mod year 100))
           (zero? (mod year 400)))))

(def weekdays [:monday :tuesday :wednesday :thursday :friday :saturday :sunday])

(defn- month->map [year month]
  (let [local-date (YearMonth/of year month)
        month (.getMonth local-date)
        days (.length month (leap? year))
        start-with (.ordinal (.getDayOfWeek (.atDay local-date 1)))]
    (reduce (fn [acc [day weekday]]
              (update acc weekday #(conj % day)))
            (zipmap weekdays (repeat (sorted-set)))
            (zipmap (range 1 (inc days)) (drop start-with (cycle weekdays))))))

(def clauses {
               :first #(first %)
               :second #(second %)
               :third #(first (drop 2 %))
               :fourth #(first (drop 3 %))
               :teenth (fn [s] (first (drop-while #(< % 13) s)))
               :last #(last %)})

(defn get-day-by-clause [month weekday clause]
  ((clause clauses) (month weekday)))

(defn meetup [month year weekday clause]
  (let [day (get-day-by-clause (month->map year month) weekday clause)]
    [year month day]))
